caddy.go

```diff
@@ -670,6 +671,85 @@ func Validate(cfg *Config) error {
 	return err
 }
 
+// ExportStorage calls List() and Load() on the configured
+// storage module and returns the underlying assets. 
+func ExportStorage(cfg *Config) (files []struct{
+	Name string
+	Body []byte
```

Comment:
Loading the entire storage into memory is probably not ideal -- it will blow up for large deployments (where a feature like this is most useful).

Maybe exporting should write directly [to an archive](https://github.com/mholt/archiver).

Although, we're getting into separate tooling territory.

What if, instead, we had a `transfer storage` utility that simply moved the contents of one storage to another storage? We could do it kv-pair by kv-pair, much more efficiently, directly, without any need to read everything into memory or have separate import/export ops.

Reply:
> What if, instead, we had a transfer storage utility that simply moved the contents of one storage to another storage? We could do it kv-pair by kv-pair, much more efficiently, directly, without any need to read everything into memory or have separate import/export ops.

Yeah that's a good point, like just Load() from old -> Store() in new -> Delete() from old per each key at a time.

Reply:
Yes! Errrr, but maybe make Delete() opt-in, and maybe that should only be done after the migration is 100% totally completed successfully.

Reply:
Transfer alone might not be enough in situations where you want to move from one cluster to another where the storage is completely isolated in the old vs new. I think adding transfer is a good idea, but I don't think it should be instead of export/import, I think both should exist.

Reply:
I don't think backups are in the scope of the web server. I'd like to leave that up to backup utilities.

I also would like to hear of actual use cases that require an export but not an import -- AFAIK the need is to be able to switch storage providers, migrate certs from one place to another. A plain export can be done with the storage system itself (i.e. copy the files from disk, or run a SQL query) -- importing so that Caddy can use it is harder, and it makes sense to be able to import from any data source.

The more I think about this the more I'm convinced that we need a transfer utility -- effectively an "import", but of course supporting any storage module as a source -- not an exporter.

Reply:
Like I said, if the storage is isolated in such a way that Caddy cannot contact both locations, then transfers cannot work. I don't think we can presume that transfers are always possible. That would be a very unfortunate limitation.

And for backups, that assumes the users know how to perform backups using their storage mechanism, and that they might even care to backup anything other than Caddy's data in that storage. For example for Redis, I might be using it as a cache otherwise, so I don't care to backup anything else.

Also re memory usage of loading everything in storage, I don't think that's a relevant concern. The CLI script is what will be keeping it in memory. Each LE cert+key+json is roughly 6k in size (with default key type etc), so 1000 certs (which is quite large) is ~6MB, and a huge deployment of like 50,000 certs would be like ~300MB; on a deployment that large, they'd surely be using a machine with plenty of RAM, hopefully over-provisioned.

Either way, writing straight to a mutable archive file is a good idea but I wouldn't consider that absolutely necessary.

Reply:
> Also re memory usage of loading everything in storage, I don't think that's a relevant concern.

fwiw this was my first impression as well -- that everything in storage is relatively small both individually and overall, and that the storage is strictly an interface for items critical to operation rather than anything served by itself ("must not be a cache..." note in the docs). I'm definitely not experienced with any large deployments though.

Reply:
> A plain export can be done with the storage system itself (i.e. copy the files from disk, or run a SQL query) -- importing so that Caddy can use it is harder, and it makes sense to be able to import from any data source.

I agree that exporting already has some clear ways to do it. The framing that makes sense to me is that commands as originally described seem useful to help _facilitate_ transfers/backups (either manually or scripted), but actually doing those would indeed be out of scope. So it's a question of whether it's too little to justify adding rather than too much. 

> The more I think about this the more I'm convinced that we need a transfer utility -- effectively an "import", but of course supporting any storage module as a source -- not an exporter.

I don't follow quite as much with this. I don't want to muddle things too much, I guess my thought for now is any potential memory issue seems orthogonal to whatever CLI semantics are desired.

Reply:
I feel like it's just bad form though -- even if the data happens to only be a few MB in most cases, we do know there are clusters out there with 100,000+ certs. It's so easy to stream the contents file-by-file, so we should probably do that. If someone comes to us complaining that there's a lot of memory use when doing an import/export (because, memory is a shared resource), then I would have no good excuse for that.

Reply:
Which is why I'm saying it makes sense to have both import/export _and_ transfer. It's up to the user to pick which is best for them. But only having transport and not import/export isn't acceptable IMO.